spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoSchemaEvolutionSuite.scala

```diff
@@ -67,54 +67,47 @@ trait MergeIntoSchemaEvolutionMixin {
-            }
-            errorContains(ex.getMessage, expectErrorWithoutEvolutionContains)
-          } else {
-            executeMerge(s"delta.`$tempPath` t", s"source s", cond,
-              clauses.toSeq: _*)
-            checkAnswer(
-              spark.read.format("delta").load(tempPath),
-              expectedWithoutEvolution.collect())
-            assert(
-              spark.read.format("delta").load(tempPath).schema.asNullable ===
-                expectedWithoutEvolution.schema.asNullable)
+    def executeMergeAndAssert(
+        df: DataFrame,
+        error: String,
+        // Ideally we would like to use `MergeClause*` instead of `Seq[MergeClause]` but it's not
+        // allowed in Scala 2.13. The error message is:
+        //    Repeated parameters are only allowed in method signatures. Use `Seq` instead.
+        executeMergeFunction: (String, String, String, Seq[MergeClause]) => Unit): Unit = {
+      append(targetData, partitionCols)
+      withTempView("source") {
+        sourceData.createOrReplaceTempView("source")
+
+        if (error != null) {
+          val ex = intercept[AnalysisException] {
+            executeMergeFunction(s"delta.`$tempPath` t", "source s", cond, clauses.toSeq)
           }
+          errorContains(Utils.exceptionString(ex), error)
+        } else {
+          executeMergeFunction(s"delta.`$tempPath` t", "source s", cond, clauses.toSeq)
+          checkAnswer(spark.read.format("delta").load(tempPath), df.collect())
+          assert(spark.read.format("delta").load(tempPath).schema.asNullable ===
+            df.schema.asNullable)
         }
       }
     }
 
-    test(s"schema evolution - $name") {
-      withSQLConf((confs :+ (DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key, "true")): _*) {
-        append(targetData, partitionCols)
-        withTempView("source") {
-          sourceData.createOrReplaceTempView("source")
+    test(s"schema evolution - $name - with evolution disabled") {
+      withSQLConf(confs: _*) {
+        executeMergeAndAssert(
+          expectedWithoutEvolution,
+          expectErrorWithoutEvolutionContains,
+          (tgt, src, cond, clauses) => executeMerge(tgt, src, cond, clauses: _*))
+      }
+    }
 
-          if (expectErrorContains != null) {
-            val ex = intercept[AnalysisException] {
-              executeMerge(s"delta.`$tempPath` t", s"source s", cond,
-                clauses.toSeq: _*)
-            }
-            errorContains(Utils.exceptionString(ex), expectErrorContains)
-          } else {
-            executeMerge(s"delta.`$tempPath` t", s"source s", cond,
-              clauses.toSeq: _*)
-            checkAnswer(
-              spark.read.format("delta").load(tempPath),
-              expected.collect())
-            assert(spark.read.format("delta").load(tempPath).schema.asNullable ===
-              expected.schema.asNullable)
-          }
-        }
+    test(s"schema evolution - $name - on via DeltaSQLConf") {
```

Comment:
If I'm reading between the lines, `on via DeltaSQLConf` may suggest we'll run the tests via the SQL/scala API to enable schema evolution?
There's a (very) large number of schema evolution and for most/all of them it doesn't really matter how schema evolution is enabled so we should just have a few tests covering enabling schema evolution and all existing run with the existing sql conf.

Reply:
Make sense. That will actually simplify the change a lot as most changes in this file are to prepare for plugging the SQL syntax into existing tests.

Reply:
Fixed!